Lib/asyncio/locks.py

```diff
@@ -181,18 +181,22 @@ def locked(self):
             self._locked = True
             return True
 
+        _waiters = self._waiters
         fut = self._loop.create_future()
-        self._waiters.append(fut)
+        _waiters.append(fut)
         try:
-            await fut
-            self._locked = True
-            return True
+            try:
+                await fut
+            finally:
+                _waiters.remove(fut)
         except futures.CancelledError:
```

Comment:
@1st1 @asvetlov I'll be honest, the only problem I have (which originates from the original code anyway) is what will happen if the exception isn't a CancelledError (such as `task.__step(self, exc=OutOfCoconutsException)`). Apart from Monty Python having no horse, the code will enter a deadlock as no one will wake up the next waiter. I'm not entirely sure what exceptions can pass into the task's __step as task doesn't allow set_exception but still better be safe than sorry no?